python/ray/data/impl/output_buffer.py

```diff
@@ -0,0 +1,71 @@
+from typing import Callable, Any, Optional
+
+from ray.data.block import Block
+from ray.data.impl.arrow_block import DelegatingArrowBlockBuilder
+
+
+class BlockOutputBuffer(object):
+    """Generates output blocks of a given size given a stream of inputs.
+
+    This class is used to turn a stream of items / blocks of arbitrary size
+    into a stream of blocks of ``target_max_block_size``. The caller should
+    check ``has_next()`` after each ``add()`` call, and call ``next()`` to get
+    the next block when ``has_next()`` returns True.
+
+    When all items have been added, the caller must call ``finalize()`` and
+    then check ``has_next()`` one last time.
+
+    Examples:
+        >>> # Yield a stream of output blocks.
+        >>> output = BlockOutputBuffer(udf, 500 * 1024 * 1024)
+        >>> for item in generator():
+        ...     output.add(item)
+        ...      if output.has_next():
+        ...         yield output.next()
+        ... output.finalize()
+        ... if output.has_next()
+        ...    yield output.next()
+    """
+
+    def __init__(self, block_udf: Optional[Callable[[Block], Block]],
+                 target_max_block_size: int):
+        from ray.data.impl.arrow_block import DelegatingArrowBlockBuilder
+        self._target_max_block_size = target_max_block_size
+        self._block_udf = block_udf
+        self._buffer = DelegatingArrowBlockBuilder()
+        self._returned_at_least_one_block = False
+        self._finalized = False
+
+    def add(self, item: Any) -> None:
+        """Add a single item to this output buffer."""
+        assert not self._finalized
+        self._buffer.add(item)
+
+    def add_block(self, block: Block) -> None:
+        """Add a data block to this output buffer."""
+        assert not self._finalized
+        self._buffer.add_block(block)
+
+    def finalize(self) -> None:
+        """Must be called once all items have been added."""
+        assert not self._finalized
+        self._finalized = True
+
+    def has_next(self) -> bool:
+        """Returns true when a complete output block is produced."""
+        if self._finalized:
+            return not self._returned_at_least_one_block \
+                or self._buffer.num_rows() > 0
+        else:
+            return self._buffer.get_estimated_memory_usage() > \
+                self._target_max_block_size
+
+    def next(self) -> Block:
+        """Returns the next complete output block."""
+        assert self.has_next()
+        block = self._buffer.build()
```

Comment:
One follow-up item that would be nice is to make the block output buffer resilient to overallocation, especially for large added blocks, where it will split a block at (or near/under) the target max block size. Right now, if you added 2 450 MiB blocks to this buffer, it would output a 900 MiB block instead of a 500 MiB while retaining a 400 MiB buffer.

Reply:
Ideally we could share the implementation with the `Batcher`, which already implements such chunking logic: https://github.com/ray-project/ray/blob/799c9bfb625931e61ec0e97f28c0836197ede48a/python/ray/data/impl/batcher.py

Reply:
Yep for sure. It's a little tricky due to the way SizeEstimator is currently written. Alternatively we can reduce the threshold; it's not super important blocks are even as long as they're bounded in size.